### DISCUSSION TITLE
Cycle detected in outlines

### DISCUSSION THREAD

**Discussion by k4fka** [pymupdf/PyMuPDF#2423] [Q&A]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)
"Cycle detected in outlines"  This might not be a bug but an issue with my own PDF.  After deleting some bookmarks I no longer see this error.   I am looking for some insight into how to identify where this cycle is coming from.

I get this warning when I run:
fitz.TOOLS.mupdf_warnings()

## To Reproduce (mandatory)
I am looking for some insight into how to identify where this cycle is coming from.

## Expected behavior (optional)
I am trying to generate a table of contents. When I call doc.get_toc() and the warning  is present. The get_toc() returns nothing. 

## Your configuration (mandatory)
 - macOS
Python 3.11.3 (main, Apr  7 2023, 19:25:52) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin - PyMuPDF version, installation 

Bad or missing prev pointer in outline tree, repairing
... repeated 2 times...
Cycle detected in outlines
[]

**Comment by JorjMcKie** [comment]
Your assumption is correct, so I took the liberty to move your post to "Discussions".

Your PDF contains errors: The table of contents (TOC) consists of objects with a lot of forward and backward pointers. Obviously, there were an endless loop ("cycle") and invalid backward ("prev") pointers detected within those objects.

**Reply by k4fka** [reply]
Thank you so much for taking the time to respond.  Out of curiosity, is there any way to retrieve a list of these outlines and their respective references?  I've been trying to  understand what bookmark is causing this issue.

**Reply by JorjMcKie** [reply]
> Thank you so much for taking the time to respond. Out of curiosity, is there any way to retrieve a list of these outlines and their respective references? I've been trying to understand what bookmark is causing this issue.

You are welcome - that is what this platform is for 😎.

When does the error occur - at fitz.open time? If yes, it might be too late for investigation, because the base library may already be done with its repair - blurring the original configuration.

But we have low-level functions for native PDF source access. So it might still be worthwhile trying these simple things:
```python
In [2]: doc=fitz.open("pymupdf.pdf")
In [3]: cat = doc.pdf_catalog()
In [4]: print(doc.xref_object(cat))
<<
  /Type /Catalog
  /Pages 10055 0 R
  /Outlines 10056 0 R
  /Names 10433 0 R
  /PageMode /UseOutlines
  /PageLabels <<
    /Nums [ 0 <<
          /S /D
        >> 2 <<
          /S /r
        >> 4 <<
          /S /D
        >> ]
  >>
  /OpenAction 1148 0 R
>>
In [5]: print(doc.xref_object(10056))  # outline start object
<<
  /Type /Outlines
  /First 5 0 R
  /Last 1146 0 R
  /Count 30
>>
In [6]: print(doc.xref_object(5))  # /First pointer
<<
  /Title 6 0 R
  /A 2 0 R
  /Parent 10056 0 R
  /Next 9 0 R
>>
In [7]: print(doc.xref_object(1146))  # /Last point
<<
  /Title 1147 0 R
  /A 1144 0 R
  /Parent 10056 0 R
  /Prev 1142 0 R
>>
In [9]: print(doc.xref_object(9))  # a /Next pointer
<<
  /Title 10 0 R
  /A 7 0 R
  /Parent 10056 0 R
  /Prev 5 0 R
  /Next 13 0 R
>>
```
In addition, there exist `/Down` pointers.
If the repair already has spoiled the structure, you can still look at the PDF surce with a suitable editor. Especially at / around the bookmarks that removed your problem upon deletion.
